File: Leetcode/787.go
func findCheapestPrice(n int, flights [][]int, src int, dst int, k int) int {
    const INF int = 0x3f3f3f3f
    var dist = make([]int, n)
    for i := 0; i < n; i++ {
        dist[i] = INF
    }
    dist[src] = 0
    var u, v, w int
    var last = make([]int, n)
    for i := 0; i < k+1; i++ {
        copy(last, dist)
        for _, j := range flights {
            u, v, w = j[0], j[1], j[2]
            if last[u] + w < dist[v] {
                dist[v] = last[u] + w
            }
        }
    }
    if dist[dst] >= INF {
        return -1
    }
    return dist[dst]
}

File: Leetcode/787/787.cpp
constexpr int INF = 0x3f3f3f3f;

class Solution {
public:
    int bellmanford(const vector<vector<int>>& edges, int n, int src, int dst, int k) {
        vector<int> dist(n, INF);
        dist[src] = 0;
        int u, v, w;
        for (int i = 0; i < k+1; ++i) {
            vector<int> last = dist;
            for (const auto &j : edges) {
                u = j[0]; v = j[1]; w = j[2];
                // relax
                if (dist[v] > last[u] + w) {
                    dist[v] = last[u] + w;
                }
            }
        }
        return (dist[dst] >= INF) ? -1 : dist[dst]; // no negative weights in the graph
    }
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        return bellmanford(flights, n, src, dst, k);
    }
};